### ISSUE TITLE
Feature: Implement State-Based Neural Trait Embeddings for Genome IDs

### ISSUE THREAD

**Issue by csmangum** [Dooders/AgentFarm#91]
Create a neural embedding system that encodes an agent's state into a fixed-length trait hash vector. This vector will be part of the genome ID, allowing direct comparison of agent behaviors and characteristics through their state representations.

## Objectives
- Create a state encoder that converts agent state into fixed-dimension embeddings
- Use this embedding as the trait hash in genome IDs
- Maintain backwards compatibility with string-based trait hashes
- Provide utilities for state comparison and behavioral analysis

#### StateEncoder Class
```python
class StateEncoder(nn.Module):
    """Encodes agent state into a fixed-dimension embedding."""
    def __init__(self, 
                 state_dims: Dict[str, int],  # e.g., {"position": 2, "velocity": 2, ...}
                 embedding_dim: int = 32,
                 hash_length: int = 6):  # for compatibility with string hashes
        super().__init__()
        
        # Separate encoders for different state types
        self.encoders = nn.ModuleDict({
            "continuous": nn.Linear(state_dims["continuous"], 16),
            "discrete": nn.Embedding(state_dims["discrete"], 16),
            "categorical": nn.Embedding(state_dims["categorical"], 16)
        })
        
        # Combine into final embedding
        self.fc = nn.Sequential(
            nn.Linear(48, embedding_dim),  # 16 * 3 = 48
            nn.ReLU(),
            nn.Linear(embedding_dim, embedding_dim)
        )
        
        # Optional: Project to hash space
        self.to_hash = nn.Linear(embedding_dim, hash_length * 4)  # 4 bits per hash char
```

#### Enhanced AgentState Class
```python
class AgentState(BaseModel):
    """Represents the complete state of an agent."""
    continuous_values: Dict[str, float]  # e.g., position, velocity
    discrete_values: Dict[str, int]      # e.g., health, energy
    categorical_values: Dict[str, str]    # e.g., action, status
    
    def to_embedding(self, encoder: StateEncoder) -> torch.Tensor:
        """Convert state to embedding vector."""
        
    def to_hash(self, encoder: StateEncoder) -> str:
        """Convert state embedding to hex hash."""
```

#### Phase 1: State Representation
- [ ] Define comprehensive agent state structure
- [ ] Create state normalization utilities
- [ ] Implement state validation and preprocessing
- [ ] Add state serialization/deserialization

#### Phase 2: Encoder Architecture
- [ ] Implement StateEncoder
- [ ] Create training infrastructure
- [ ] Add vector-to-hash conversion
- [ ] Implement hash-to-vector approximation

#### Phase 3: Integration with GenomeId
- [ ] Update GenomeId to support vector trait hashes
- [ ] Add backwards compatibility layer
- [ ] Create migration utilities
- [ ] Update database schema

#### Phase 4: Analysis Tools
- [ ] Add state comparison utilities
- [ ] Create behavioral clustering tools
- [ ] Implement trajectory analysis
- [ ] Add visualization tools

### Dependencies to Add
```toml
torch>=2.0.0
numpy>=1.20.0
scikit-learn>=1.0.0
```

### New Files Needed
```
state_embedding/
├── __init__.py
├── state.py           # AgentState implementation
├── encoder.py         # StateEncoder implementation
├── training.py        # Training infrastructure
├── analysis.py        # State analysis tools
└── utils.py          # Utility functions
```

### Example Usage
```python
# Define agent state
state = AgentState(
    continuous_values={
        "position_x": 0.5,
        "position_y": 0.3,
        "velocity": 0.8
    },
    discrete_values={
        "health": 100,
        "energy": 50
    },
    categorical_values={
        "action": "move_forward",
        "status": "active"
    }
)

# Create encoder
encoder = StateEncoder(
    state_dims={
        "continuous": 3,  # position_x, position_y, velocity
        "discrete": 2,    # health, energy
        "categorical": 2  # action, status
    }
)

# Generate trait hash
trait_hash = state.to_hash(encoder)  # returns hex string
trait_vector = state.to_embedding(encoder)  # returns tensor
```

### Loss Functions
1. **State Continuity Loss**: Similar states should have similar embeddings
2. **Temporal Consistency Loss**: Sequential states should have smooth embedding transitions
3. **Behavioral Clustering Loss**: States with similar behaviors should cluster together

### Training Data
- Collect state sequences from agent interactions
- Use state transitions for temporal consistency
- Group similar behaviors for clustering

## Acceptance Criteria
1. State encoder can handle all agent state types
2. Trait hash maintains compatibility with existing system
3. Similar states produce similar embeddings (verified through clustering)
4. Can reconstruct approximate state from embedding
5. Performance benchmarks:
   - Encoding time < 10ms
   - Hash generation < 5ms
   - State comparison < 1ms

## Future Enhancements
- Add support for dynamic state dimensions
- Implement hierarchical state encoding
- Add reinforcement learning integration
- Create behavior prediction models
